win32/src/file_win32.c

```diff
@@ -60,4 +41,472 @@ static void on_file_io_complete_win32(
+        (execution_engine == NULL) ||
+        /*Codes_SRS_FILE_43_002: [ If full_file_name is NULL then file_create shall fail and return NULL. ]*/
+        /*Codes_SRS_FILE_WIN32_43_048: [ If full_file_name is NULL then file_create shall fail and return NULL. ]*/
+        (full_file_name == NULL) ||
+        /*Codes_SRS_FILE_43_037: [ If full_file_name is an empty string, file_create shall fail and return NULL. ]*/
+        /*Codes_SRS_FILE_WIN32_43_059: [ If full_file_name is an empty string, file_create shall fail and return NULL. ]*/
+        (strlen(full_file_name) == 0)
+    )
+    {
+        LogError("Invalid arguments to file_create: EXECUTION_ENGINE_HANDLE execution_engine=%p, const char* full_file_name=%s, FILE_REPORT_FAULT user_report_callback=%p, void* user_report_faul_context=%p",
+            execution_engine, full_file_name, user_report_fault_callback, user_report_fault_context);
+        result = false;
+    }
+    else
+    {
+        /*Codes_SRS_FILE_WIN32_43_041: [ file_create shall allocate a FILE_HANDLE. ]*/
+        handle = malloc(sizeof(FILE_HANDLE_DATA));
+        if (handle == NULL) {
+            LogError("Failure in malloc");
+            result = false;
+        }
+        else
+        {
+            /*Codes_SRS_FILE_43_003: [ If a file with name full_file_name does not exist, file_create shall create a file with that name.]*/
+            /*Codes_SRS_FILE_43_001: [ file_create shall open the file named full_file_name for asynchronous operations and return its handle. ]*/
+            /*Codes_SRS_FILE_WIN32_43_001: [ file_create shall call CreateFileA with full_file_name as lpFileName, GENERIC_READ|GENERIC_WRITE as dwDesiredAccess, FILE_SHARED_READ as dwShareMode, NULL as lpSecurityAttributes, OPEN_ALWAYS as dwCreationDisposition, FILE_FLAG_OVERLAPPED|FILE_FLAG_WRITE_THROUGH as dwFlagsAndAttributes and NULL as hTemplateFile. ]*/
+            if ((handle->h_file = CreateFileA(
+                full_file_name,                                       /*_In_     LPCTSTR               lpFileName*/
+                GENERIC_READ | GENERIC_WRITE,                       /*_In_     DWORD                 dwDesiredAccess*/
+                FILE_SHARE_READ,                                    /*_In_     DWORD                 dwShareMode*/
+                NULL,                                               /*_In_opt_ LPSECURITY_ATTRIBUTES lpSecurityAttributes*/
+                OPEN_ALWAYS,                                        /*_In_     DWORD                 dwCreationDisposition*/
+                FILE_FLAG_OVERLAPPED | FILE_FLAG_WRITE_THROUGH,     /*_In_     DWORD                 dwFlagsAndAttributes*/
+                NULL                                                /*_In_opt_ HANDLE                hTemplateFile*/
+            )) == INVALID_HANDLE_VALUE)
+            {
+                LogError("Failure in CreateFileA, full_file_name=%s", full_file_name);

```

Comment:
>Failure [](start = 26, length = 7)

Also should print GetLastError

Reply:
This goes for any win32 function that says you should call GetLastError when it fails (so here and everywhere)

---
In reply to: [456706697](https://github.com/Azure/azure-c-pal/pull/10#discussion_r456706697) [](ancestors = 456706697)